subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradleModuleMetadataParser.java

```diff
@@ -641,6 +644,33 @@ private static class ModuleDependency {
             this.endorsing = endorsing;
             this.artifact = artifact;
         }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
💭 Since it's important that the `ModuleDependency` and `ModuleDependencyConstraint` types continue to properly identify duplicates, so they can be excluded from the set, is it worth also building unit tests to ensure proper equals/hashcode behavior?